warehouse/accounts/auth_policy.py

```diff
@@ -10,15 +10,110 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import struct
+
+from pymacaroons import Macaroon, Verifier
+from pymacaroons.caveat_delegates import ThirdPartyCaveatVerifierDelegate
+from pymacaroons.exceptions import MacaroonException
+
 from pyramid.authentication import (
     BasicAuthAuthenticationPolicy as _BasicAuthAuthenticationPolicy,
+    CallbackAuthenticationPolicy as _CallbackAuthenticationPolicy,
     SessionAuthenticationPolicy as _SessionAuthenticationPolicy,
 )
 
 from warehouse.accounts.interfaces import IUserService
 from warehouse.cache.http import add_vary_callback
 
 
+class AccountTokenAuthenticationPolicy(_CallbackAuthenticationPolicy):
+    def __init__(self, authenticate):
+        self._authenticate = authenticate
+        self.callback = self._auth_callback
+
+        self._routes_allowed = ["forklift.legacy.file_upload"]
+
+    def unauthenticated_userid(self, request):
+        if request.matched_route.name not in self._routes_allowed:
+            return None
+
+        account_token = request.params.get("account_token")
+
+        if account_token is None:
+            return None
+
+        try:
+            macaroon = Macaroon.deserialize(account_token)
+
+            # First, check identifier and location
+            if macaroon.identifier != request.registry.settings["account_token.id"]:
+                return None
+
+            if macaroon.location != "pypi.org":
+                return None
+
+            # Check the macaroon against our configuration
+            verifier = Verifier()
+
+            verifier.third_party_caveat_verifier_delegate = IgnoreThirdPartyCaveats()
```

Comment:
Ah, the name I chose here might be confusing. `IgnoreThirdPartyCaveats` just returns `True` for verifying any third-party caveat (since we don't know how those caveats are structured). This does allow macaroons to include third-party caveats, which we ignore (in the sense that they will not cause validation to fail).